### PR TITLE
Add alternative LRUCache version

### PR DESCRIPTION
**Pull request by lgbo-ustc**
Changelog category (leave one):
- Not for changelog (changelog entry is not required)

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add evict policy support in LRUCache which wold be used in #31104

Detailed description / Documentation draft:

Add LRUCache implementation which can have unreleasable keys and manages their release state itself.

### CODE REVIEW

File: src/Common/LRUCache.h
Comment (kssenii) [inline src/Common/LRUCache.h:10]: unused header

File: src/Common/LRUCache.h
Comment (kssenii) [inline src/Common/LRUCache.h:26]: I think we can remove this comment (and the same below), because it is more like implementation detail you used.

File: src/Common/LRUCache.h
Comment (kssenii) [inline src/Common/LRUCache.h:160]: @alexey-milovidov I don't know what to do with these two almost identical variables. The difference is that previously if we put key in cache it could be removed just after we did that by calling `removeOverflow()` at the end of `setImpl`. And `is_value_loaded` could  mean that it was loaded but removed from cache right after. Therefore now appears another variable (`is_value_updated`) to check that we loaded and actually managed to put it.
I would rather have one variable instead of two almost identical, but this would change previous behaviour of `getOrSet` method a bit.

File: src/Common/LRUCache.h
Comment (kssenii) [inline src/Common/LRUCache.h:24]: typo
```suggestion
struct TrivialLRUCacheEvictPolicy
```

File: src/Common/LRUCache.h
Comment (kssenii) [inline src/Common/LRUCache.h:112]: Better to return result as struct
```
struct Result
{
    MappedPtr value;
    bool cache_miss = true;
    bool set_successful = false;
}
```

File: src/Common/LRUCache.h
Comment (kssenii) [reply src/Common/LRUCache.h:160]: Resolved, not possible to unite.

File: src/Common/LRUCache.h
Comment (kssenii) [inline src/Common/LRUCache.h:348]: I would rather remove such comments.

File: src/Common/LRUCache.h
Comment (kssenii) [reply src/Common/LRUCache.h:112]: Same for `getOrSet`. But with comment for `set_successful` that for `getOrSet` it is not trustworthy because `removeOverflow` is called right after set.

Comment (lgbo-ustc) [reply src/Common/LRUCache.h]: Do you mean also using Result as the return of `getOrSet()`?

Comment (kssenii) [reply src/Common/LRUCache.h]: Yes, for consistency, because it will look strange that `getOrSet` returns `std::pair` and `getOrTrySet` returns `struct`.

Comment (kssenii) [reply src/Common/LRUCache.h]: @lgbo-ustc  Actually lets remove `getOrTrySet` at all. As I remember we don't really need it.

Comment (lgbo-ustc) [reply src/Common/LRUCache.h]: If so, maybe `getOrTrySet()` could be deleted

Comment (kssenii) [reply src/Common/LRUCache.h]: Yes, I just wrote it above.

Comment (lgbo-ustc) [reply src/Common/LRUCache.h]: done

File: src/Common/LRUCache.h
Comment (kssenii) [inline src/Common/LRUCache.h:151]:
```suggestion
        bool is_value_loaded_and_set = false;
```

File: src/Common/LRUCache.h
Comment (kssenii) [inline src/Common/LRUCache.h:371]:
```suggestion
            if (!evict_policy.canRelease(cell.value))
                return false;
            evict_policy.release(cell.value);
```

File: src/Common/LRUCache.h
Comment (kssenii) [inline src/Common/LRUCache.h:380]: Looks incorrect. In this case if there already was a value with the same key in `cells` (case `!inserted`), then when we update its value, it can have a different size (because we change previous value to `Mapped`), so there can be overflow, but now there is no `removeOverflow` at the end of this method.

**Review by kssenii — CHANGES_REQUESTED** [review]

File: src/Common/LRUCache.h
Comment (lgbo-ustc) [reply src/Common/LRUCache.h:378]: updated

File: src/Common/LRUCache.h
Comment (kssenii) [inline src/Common/LRUCache.h:355]: Also this changes previous behaviour in case in try catch below we get an exception. I mean now we could remove some values with `removeOverflow`, then get exception and not insert anything, but we already made place for this insertion. (Even though this is an exception from insertion into list and not in general likely to happen). So better to first insert into queue, remember iterator to currently inserted key, then check if there is enough space, it not -- remove from queue.

Comment (lgbo-ustc) [reply src/Common/LRUCache.h]: done

File: src/Common/LRUCache.h
Comment (kssenii) [inline src/Common/LRUCache.h:157]:
```suggestion
            is_value_loaded_and_set = setImpl(key, token->value, cache_lock);
```

File: src/Common/LRUCache.h
Comment (kssenii) [reply src/Common/LRUCache.h]: It was a bad idea actually. Or now we need to guarantee that for key, which we want to set, `canRelease` must return true, and also one more problem -- `release()` (is it legal to call it if in fact we did not return from `set`?). For cache it removes files from local fs, and BTW does not currently check if path exists before doing that.

I'll think what to do with that, how to do it differently. If you also have ideas -- propose.
Also will need to add unit tests for changes from this pr.

Comment (kssenii) [reply src/Common/LRUCache.h]: Ok no problem here after all because value will be `nullptr`, but then need to move check for nullptr out of `release()` and `canRelease()` and pass a reference inside.

File: src/Common/LRUCache.h
Comment (kssenii) [inline src/Common/LRUCache.h:368]: Also I am worried that `release` can throw exception...
But not sure should we catch it here or just make sure in `release` method that it never throws.

File: src/Common/LRUCache.h
Comment (lgbo-ustc) [reply src/Common/LRUCache.h]: btw, I don't find where to add a unit test

File: src/Common/LRUCache.h
Comment (kssenii) [reply src/Common/LRUCache.h:368]: Ok I the only solution is to make `release` and `canRelease` as `noexcept`. Because it is controversial if or how we should process exceptions from `release` for old keys.

File: src/Common/LRUCache.h
Comment (kssenii) [reply src/Common/LRUCache.h]: Similar to `src/Common/tests/gtest_lru_hash_map.cpp`
to run:
```
 ninja unit_tests_dbms
 cd build/src
./unit_tests_dbms --gtest_filter='TransformQueryForExternalDatabase*' --gtest_brief=0 --gtest_throw_on_failure=1 
```

Comment (lgbo-ustc) [reply src/Common/LRUCache.h]: Another fatal bug, when `canRelease()` returns true, before calling `release()`, the element may turn into unreleasable again

Comment (kssenii) [reply src/Common/LRUCache.h]: > Another fatal bug, when canRelease() returns true, before calling release(), the element may turn into unreleasable again

As I understood you, what you say should not happen because we hold mutex in `set` and in `get`. It may become unreleasable if someone successfully called `get`, but while we are in `trySet` this will not happen.

Comment (lgbo-ustc) [reply src/Common/LRUCache.h]: following is the case
1. we get A from cache, A is releasable
2. then we try to set B and sees A is releasable, 
3. at now something happpens on A outside lru cache, and make A unreleasable
4. then lru cache call release A to make space for B.

Comment (kssenii) [reply src/Common/LRUCache.h]: This is the case I asked you to fix here https://github.com/ClickHouse/ClickHouse/pull/31104#discussion_r775749309.
So now I believe that this is case is handled by your code of cache for storage because getting key from cache and creating a value holder object is done under mutex. And getting/setting keys in cache always goes throw that mutex.

Comment (lgbo-ustc) [reply src/Common/LRUCache.h]: In our use case, there is and outside mutex which could avoid this case, but we cannot gurantee for other use cases

Comment (kssenii) [reply src/Common/LRUCache.h]: Ok I agree, then may be instead of `bool canRelease` and `void release` have just `bool tryRelease()`?

Comment (lgbo-ustc) [reply src/Common/LRUCache.h]: I just thought use `markRelease` for mark the element could be used anymore and `unmarkRelease` for rollback `removeOverflow`, but this may still have side effect

Comment (kssenii) [reply src/Common/LRUCache.h]: When would we need to rollback it?

Comment (lgbo-ustc) [reply src/Common/LRUCache.h]: It seems more likely what we need is a resource pool with key index and lru evict policy. The apis may are as following
```
// if could not insert into the cache, return nullptr
// if result is not null,  increase its reference
// use lru evict policy inside to evict elements which reference is 0
std::shared_ptr<Value> acquire(const Key & key,  LoadFunc && load_func);
// decrease its reference
void release(const Key & key, std::shared_ptr<Value> v);
```

Comment (lgbo-ustc) [reply src/Common/LRUCache.h]: > When would we need to rollback it? May be just make implementation when we would not need rollback?

for example, we have elements : a,b,c, and a is releasabel, other two not. Now we want to insert d, but we cannot find enough space for d even after evict a. Maybe we just want to keep the cache elements as before. a, b and c are still in the cache

Comment (kssenii) [reply src/Common/LRUCache.h]: > I still prefer a new implementation but not LRUCache

Yes, let's try to make a new one. I also start to think it will be better.

**Comment by lgbo-ustc** [comment]
@kssenii plz see the new cache, `LRUResourceCache`

File: src/Common/LRUResourceCache.h
Comment (kssenii) [inline src/Common/LRUResourceCache.h:204]: Style. Should be `insertValue` at least. But I like name `set(...)` more (or `setImpl`).

File: src/Common/tests/gtest_lru_resource_cache.cpp
Comment (kssenii) [inline src/Common/tests/gtest_lru_resource_cache.cpp:38]: Ok but I also stared to think that it makes sense to have a method (not sure how to name it, may be just `get`). It will return a holder, which will call `acquire` in constructor and `release` in desctructor. And may be make public only this method, `acquire` and `release` may be private. What do you think?

Comment (lgbo-ustc) [reply src/Common/tests/gtest_lru_resource_cache.cpp:38]: good idea, I take a try

Comment (lgbo-ustc) [reply src/Common/tests/gtest_lru_resource_cache.cpp]: updated

File: src/Common/LRUResourceCache.h
Comment (kssenii) [inline src/Common/LRUResourceCache.h:43]: Why would we need such method? It seems redundant to me.

File: src/Common/LRUResourceCache.h
Comment (kssenii) [inline src/Common/LRUResourceCache.h:42]:
```suggestion
        Mapped & value() { return *value; }
```

File: src/Common/LRUResourceCache.h
Comment (kssenii) [inline src/Common/LRUResourceCache.h:64]:
```suggestion
        auto mapped_ptr = getImpl(key);
```

File: src/Common/LRUResourceCache.h
Comment (lgbo-ustc) [reply src/Common/LRUResourceCache.h:43]: In our case, when a remote file is updated, we need to drop the current cached file

Comment (kssenii) [reply src/Common/LRUResourceCache.h:43]: Ok, but then may be it makes sense to make it's implementation such that if refcount is 0, it removes the key, if non-zero refcount, then it sets the flag in cell `expired = true`. 
So if `expired == true`, then 
1. `get` for this key will return false
2. the weight of value if this key is still taken into account when we check remaining space
3. when `release` is called for this key and refcount gets to 0, then we remove this key from cache

Also I see no reason why it should accept a holder, and not just key. If we have a holder and then decide that corresponding key needs to be removed, then better to get rid of holder first. I mean no need in such method in holder to me.

Comment (lgbo-ustc) [reply src/Common/LRUResourceCache.h]: done

**Comment by kssenii** [comment]
@lgbo-ustc can you please enable edits by maintainers (on the right menu of this PR)? So I can also push to your branch.

**Comment by lgbo-ustc** [comment]
image

There is no menu here. is it enabled when this pr is crearted?

**Comment by kssenii** [comment]
> There is no menu here. is it enabled when this pr is crearted?

hm, Ok, sad. Usually the button to allow edits just below `notifications` section. And usually when you create a pr there is also such botton, but it is true by default, i.e. usually I can push to forks without even asking to allow that manually. 
(though you are not the first who does not have this allow edits button)

File: src/Common/LRUResourceCache.h
Comment (kssenii) [inline src/Common/LRUResourceCache.h:149]: Why don't you add `InsertTokenHolder`?

Comment (lgbo-ustc) [reply src/Common/LRUResourceCache.h]: updated

File: src/Common/LRUResourceCache.h
Comment (lgbo-ustc) [reply src/Common/LRUResourceCache.h]: done

File: src/Common/LRUCache.h
Comment (kssenii) [inline src/Common/LRUCache.h:75]: Forgot to also remove from `queue`?
btw let's remove it since it is not needed now?

File: src/Common/tests/gtest_lru_resource_cache.cpp
Comment (kssenii) [inline src/Common/tests/gtest_lru_resource_cache.cpp:56]: Can also be like this
```suggestion
    holder1.reset();
```
(not necessary)

File: src/Common/LRUCache.h
Comment (lgbo-ustc) [reply src/Common/LRUCache.h:76]: fixed

File: src/Common/tests/gtest_lru_resource_cache.cpp
Comment (lgbo-ustc) [reply src/Common/tests/gtest_lru_resource_cache.cpp]: done

**Review by kssenii — APPROVED** [review]
Ok, let's wait for tests and merge.
I also need this implementation for something else so possibly it will be reworked in some way (and I will test it more).

**Pull request merged**
